raster3d/r3.info/main.c

```diff
@@ -130,368 +181,810 @@ int main(int argc, char *argv[])
-                              name);
+            if (hist_ok) {
+                printline("  Data Source:");
+                if (G_asprintf(&line, "   %s",
+                               Rast_get_history(&hist, HIST_DATSRC_1)) > 0)
+                    printline(line);
+                else
+                    G_fatal_error(_("Cannot allocate memory for string"));
+
+                if (G_asprintf(&line, "   %s",
+                               Rast_get_history(&hist, HIST_DATSRC_2)) > 0)
+                    printline(line);
+                else
+                    G_fatal_error(_("Cannot allocate memory for string"));
+
+                printline("");
+
+                printline("  Data Description:");
+                if (G_asprintf(&line, "   %s",
+                               Rast_get_history(&hist, HIST_KEYWRD)) > 0)
+                    printline(line);
+                else
+                    G_fatal_error(_("Cannot allocate memory for string"));
+
+                printline("");
+                if (Rast_history_length(&hist)) {
+                    printline("  Comments:  ");
 
-            Rast3d_range_min_max(g3map, &dmin, &dmax);
+                    for (i = 0; i < Rast_history_length(&hist); i++)
 
-            if (dmin != dmin)
-                snprintf(tmp1, sizeof(tmp1), "%s", "NULL");
-            else
-                format_double(dmin, tmp1);
-            if (dmax != dmax)
-                snprintf(tmp2, sizeof(tmp2), "%s", "NULL");
-            else
-                format_double(dmax, tmp2);
+                    /**************************************/
+                    {
+                        if (G_asprintf(&line, "   %s",
+                                       Rast_history_line(&hist, i)) > 0)
+                            printline(line);
+                        else
+                            G_fatal_error(
+                                _("Cannot allocate memory for string"));
+                    }
+                }
 
-            if (G_asprintf(&line, "  Range of data:   min = %10s max = %10s",
-                           tmp1, tmp2) > 0)
-                printline(line);
-            else
-                G_fatal_error(_("Cannot allocate memory for string"));
-        }
+                printline("");
+            }
 
-        printline("");
+            divider('+');
+
+            fprintf(out, "\n");
+            break;
+
+        case SHELL:
+            fprintf(out, "layer=%s\n", name);
+            fprintf(out, "date=%s\n",
+                    hist_ok ? Rast_get_history(&hist, HIST_MAPID) : "??");
+            fprintf(out, "mapset=%s\n", mapset);
+            fprintf(out, "creator=%s\n",
+                    hist_ok ? Rast_get_history(&hist, HIST_CREATOR) : "??");
+            fprintf(out, "project=%s\n", G_location());
+            fprintf(out, "database=%s\n", G_gisdbase());
+            fprintf(out, "title=%s\n",
```

Comment:
Some of these should be probably with quotes. r.info has quotes, v.info does not. But quotes sounds like a good idea.